app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,379 @@
+        yCenteredDOMPos = graph.getNodePosition(START_NODE_ID).y;
+      } else {
+        const xStartNodePos = graph.getNodePosition(START_NODE_ID).x;
+        graph.moveNodeToPosition(START_NODE_ID, xStartNodePos, yCenteredDOMPos);
+
+        const xEndNodePos = graph.getNodePosition(END_NODE_ID).x;
+        graph.moveNodeToPosition(END_NODE_ID, xEndNodePos, yCenteredDOMPos);
+      }
+    });
+  }
+
+  drawGraphs() {
+    this.modifyStepNames();
+    this.stageNames.forEach((stageName, i) => {
+      this.drawStageGraph(i);
+    });
+    this.adjustGraphNodePositions();
+    this.closeNonactiveSteps();
+  }
+
+  closeNonactiveSteps() {
+    const { stageResults } = this.props;
+    this.stageNames.forEach((stageName, i) => {
+      const graph = this.graphs[i];
+      const stageData = stageResults.stages[stageName];
+      if (stageData.job_status !== "STARTED") {
+        graph.afterDrawingOnce(() => {
+          this.toggleStage(i);
+        });
+      }
+    });
+  }
+
+  drawStageGraph(index) {
+    const { nodeColor, backgroundColor, edgeColor } = this.props;
+    const container = this[`container-${index}`];
+
+    const edgeData = this.generateIntraEdgeData(index).concat(
+      this.generateInterEdgeData(index)
+    );
+    const nodeData = this.generateNodeData(index, edgeData);
+
+    const options = {
+      nodes: {
+        borderWidth: 0,
+        color: nodeColor,
+        shape: "box",
+        shapeProperties: {
+          borderRadius: 6,
+        },
+        widthConstraint: {
+          minimum: 120,
+        },
+        heightConstraint: {
+          minimum: 24,
+        },
+        font: {
+          face: "Open Sans",
+        },
+      },
+      groups: {
+        startEndNodes: {
+          widthConstraint: 8,
+          heightConstraint: 0,
+          color: backgroundColor,
+          fixed: {
+            x: true,
+            y: true,
+          },
+        },
+      },
+      edges: {
+        arrows: {
+          to: {
+            enabled: true,
+            type: "arrow",
+            scaleFactor: 0.8,
+          },
+        },
+        smooth: {
+          type: "cubicBezier",
+          roundness: 0.8,
+        },
+        color: edgeColor,
+      },
+      layout: {
+        hierarchical: {
+          direction: "LR",
+          sortMethod: "directed",
+          levelSeparation: 200,
+          parentCentralization: false,
+          blockShifting: false,
+          edgeMinimization: false,
+        },
+      },
+      physics: {
+        enabled: false,
+      },
+      interaction: {
+        zoomView: false,
+        dragView: false,
+        dragNodes: false,
+      },
+    };
+
+    const currStageGraph = new NetworkGraph(
+      container,
+      nodeData,
+      edgeData,
+      options
+    );
+    currStageGraph.minimizeWidthGivenScale(1.0);
+
+    this.graphs.push(currStageGraph);
+  }
+
+  render() {
+    const stageContainers = [];
+
+    this.stageNames.forEach((stageName, i) => {
```

Comment:
a map would be appropriate here, and avoids creating an array and pushing onto it.